### ISSUE TITLE
diff-hl-define-bitmaps: Args out of range: 0, "Width must be from 1 to 16"

### ISSUE THREAD

**Issue by mgcyung** [dgutov/diff-hl#135]
When a file is opening in Emacs 28, there is an error
`
diff-hl-define-bitmaps: Args out of range: 0, "Width must be from 1 to 16"
`
which doesn't  appear in Emacs 27.

**Comment by dgutov** [comment]
Maybe you have an older build of Emacs 27? It should have the same check added in 999d75c (Feb 22).

Anyway, does this patch help?

```diff
diff --git a/diff-hl.el b/diff-hl.el
index 056b0ea..e632389 100644
--- a/diff-hl.el
+++ b/diff-hl.el
@@ -182,13 +182,15 @@ the end position as its only argument."
       )))
 
 (defun diff-hl-maybe-define-bitmaps ()
-  (when (window-system) ;; No fringes in the console.
+  (when (and (window-system)
+             (not (zerop fringe-mode))) ;; No fringes in the console.
     (unless (fringe-bitmap-p 'diff-hl-bmp-empty)
       (diff-hl-define-bitmaps)
       (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center))))
 
 (defun diff-hl-maybe-redefine-bitmaps ()
-  (when (window-system)
+  (when (and (window-system)
+             (not (zerop fringe-mode)))
     (diff-hl-define-bitmaps)))
 
 (defvar diff-hl-spec-cache (make-hash-table :test 'equal))
```

**Comment by mgcyung** [comment]
After applied the patch it shows 
`
diff-hl-dired-mode: Wrong type argument: number-or-marker-p, nil
`
And the backtrace is as follows
`
Debugger entered--Lisp error: (wrong-type-argument number-or-marker-p nil)
  diff-hl-maybe-define-bitmaps()
  diff-hl-mode(1)
  turn-on-diff-hl-mode()
  global-diff-hl-mode-enable-in-buffers()
  run-hooks(change-major-mode-after-body-hook after-change-major-mode-hook)
  normal-mode(t)
  after-find-file(nil t)
  find-file-noselect-1(#<buffer 1.txt> "~/1.txt" nil nil "~/1.txt" (19147035 66311))
  find-file-noselect("/home/magic/1.txt" nil nil nil)
  find-file("/home/magic/1.txt")
  helm-find-file-or-marked("/home/magic/1.txt")
  helm-execute-selection-action-1()
  helm-execute-selection-action()
  helm-internal(helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil)
  apply(helm-internal (helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil))
  helm(helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil)
  apply(helm (helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil))
  helm(:sources helm-source-find-files :input "/home/magic/" :case-fold-search smart :preselect nil :ff-transformer-show-only-basename t :default nil :prompt "Find files or url: " :buffer "*helm find files*")
  helm-find-files-1("/home/magic/" nil)
  helm-find-files(nil)
  funcall-interactively(helm-find-files nil)
  call-interactively(helm-find-files nil nil)
  command-execute(helm-find-files)
`

**Comment by dgutov** [comment]
OK, thanks. What about this one?

```diff
diff --git a/diff-hl.el b/diff-hl.el
index 056b0ea..a2db778 100644
--- a/diff-hl.el
+++ b/diff-hl.el
@@ -182,13 +182,15 @@ the end position as its only argument."
       )))
 
 (defun diff-hl-maybe-define-bitmaps ()
-  (when (window-system) ;; No fringes in the console.
+  (when (and (window-system)
+             (not (eql fringe-mode 0))) ;; No fringes in the console.
     (unless (fringe-bitmap-p 'diff-hl-bmp-empty)
       (diff-hl-define-bitmaps)
       (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center))))
 
 (defun diff-hl-maybe-redefine-bitmaps ()
-  (when (window-system)
+  (when (and (window-system)
+             (not (eql fringe-mode 0)))
     (diff-hl-define-bitmaps)))
 
 (defvar diff-hl-spec-cache (make-hash-table :test 'equal))

```

**Comment by mgcyung** [comment]
It shows
```
let*: Args out of range: 0, "Width must be from 1 to 16"
```
The backtrace is
```
Debugger entered--Lisp error: (args-out-of-range 0 "Width must be from 1 to 16")
  define-fringe-bitmap(diff-hl-bmp-top [0 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5] 20 0 top)
  (let* ((scale (if (and (boundp 'text-scale-mode-amount) (numberp text-scale-mode-amount)) (expt text-scale-mode-step text-scale-mode-amount) 1)) (spacing (or (and (display-graphic-p) (default-value 'line-spacing)) 0)) (h (+ (ceiling (* (frame-char-height) scale)) (if (floatp spacing) (truncate (* (frame-char-height) spacing)) spacing))) (w (min (frame-parameter nil (intern (format "%s-fringe" diff-hl-side))) 16)) (middle (make-vector h (expt 2 (1- w)))) (ones (1- (expt 2 w))) (top (copy-sequence middle)) (bottom (copy-sequence middle)) (single (copy-sequence middle))) (aset top 0 ones) (aset bottom (1- h) ones) (aset single 0 ones) (aset single (1- h) ones) (define-fringe-bitmap 'diff-hl-bmp-top top h w 'top) (define-fringe-bitmap 'diff-hl-bmp-middle middle h w 'center) (define-fringe-bitmap 'diff-hl-bmp-bottom bottom h w 'bottom) (define-fringe-bitmap 'diff-hl-bmp-single single h w 'top) (define-fringe-bitmap 'diff-hl-bmp-i [3 3 0 3 3 3 3 3 3 3] nil 2 'center) (let* ((w2 (* (/ w 2) 2)) (w2 (if (= 0 w2) 2 w2)) (delete-row (- (expt 2 (1- w2)) 2)) (middle-pos (1- (/ w2 2))) (middle-bit (expt 2 middle-pos)) (insert-bmp (make-vector w2 (* 3 middle-bit)))) (define-fringe-bitmap 'diff-hl-bmp-delete (make-vector 2 delete-row) w2 w2) (aset insert-bmp 0 0) (aset insert-bmp middle-pos delete-row) (aset insert-bmp (1+ middle-pos) delete-row) (aset insert-bmp (1- w2) 0) (define-fringe-bitmap 'diff-hl-bmp-insert insert-bmp w2 w2)))
  diff-hl-define-bitmaps()
  (if (fringe-bitmap-p 'diff-hl-bmp-empty) nil (diff-hl-define-bitmaps) (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center))
  (progn (if (fringe-bitmap-p 'diff-hl-bmp-empty) nil (diff-hl-define-bitmaps) (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center)))
  (if (and (window-system) (not (eql fringe-mode 0))) (progn (if (fringe-bitmap-p 'diff-hl-bmp-empty) nil (diff-hl-define-bitmaps) (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center))))
  diff-hl-maybe-define-bitmaps()
  (progn (diff-hl-maybe-define-bitmaps) (add-hook 'after-save-hook 'diff-hl-update nil t) (add-hook 'after-change-functions 'diff-hl-edit nil t) (add-hook (if vc-mode 'diff-hl-mode-on-hook 'find-file-hook) 'diff-hl-update t t) (add-hook 'vc-checkin-hook 'diff-hl-update nil t) (add-hook 'after-revert-hook 'diff-hl-update nil t) (add-hook 'magit-revert-buffer-hook 'diff-hl-update nil t) (add-hook 'magit-not-reverted-hook 'diff-hl-update nil t) (add-hook 'text-scale-mode-hook 'diff-hl-maybe-redefine-bitmaps nil t))
  (if diff-hl-mode (progn (diff-hl-maybe-define-bitmaps) (add-hook 'after-save-hook 'diff-hl-update nil t) (add-hook 'after-change-functions 'diff-hl-edit nil t) (add-hook (if vc-mode 'diff-hl-mode-on-hook 'find-file-hook) 'diff-hl-update t t) (add-hook 'vc-checkin-hook 'diff-hl-update nil t) (add-hook 'after-revert-hook 'diff-hl-update nil t) (add-hook 'magit-revert-buffer-hook 'diff-hl-update nil t) (add-hook 'magit-not-reverted-hook 'diff-hl-update nil t) (add-hook 'text-scale-mode-hook 'diff-hl-maybe-redefine-bitmaps nil t)) (remove-hook 'after-save-hook 'diff-hl-update t) (remove-hook 'after-change-functions 'diff-hl-edit t) (remove-hook 'find-file-hook 'diff-hl-update t) (remove-hook 'vc-checkin-hook 'diff-hl-update t) (remove-hook 'after-revert-hook 'diff-hl-update t) (remove-hook 'magit-revert-buffer-hook 'diff-hl-update t) (remove-hook 'magit-not-reverted-hook 'diff-hl-update t) (remove-hook 'text-scale-mode-hook 'diff-hl-maybe-redefine-bitmaps t) (diff-hl-remove-overlays))
  (let ((last-message (current-message))) (setq diff-hl-mode (if (eq arg 'toggle) (not diff-hl-mode) (> (prefix-numeric-value arg) 0))) (if diff-hl-mode (progn (diff-hl-maybe-define-bitmaps) (add-hook 'after-save-hook 'diff-hl-update nil t) (add-hook 'after-change-functions 'diff-hl-edit nil t) (add-hook (if vc-mode 'diff-hl-mode-on-hook 'find-file-hook) 'diff-hl-update t t) (add-hook 'vc-checkin-hook 'diff-hl-update nil t) (add-hook 'after-revert-hook 'diff-hl-update nil t) (add-hook 'magit-revert-buffer-hook 'diff-hl-update nil t) (add-hook 'magit-not-reverted-hook 'diff-hl-update nil t) (add-hook 'text-scale-mode-hook 'diff-hl-maybe-redefine-bitmaps nil t)) (remove-hook 'after-save-hook 'diff-hl-update t) (remove-hook 'after-change-functions 'diff-hl-edit t) (remove-hook 'find-file-hook 'diff-hl-update t) (remove-hook 'vc-checkin-hook 'diff-hl-update t) (remove-hook 'after-revert-hook 'diff-hl-update t) (remove-hook 'magit-revert-buffer-hook 'diff-hl-update t) (remove-hook 'magit-not-reverted-hook 'diff-hl-update t) (remove-hook 'text-scale-mode-hook 'diff-hl-maybe-redefine-bitmaps t) (diff-hl-remove-overlays)) (run-hooks 'diff-hl-mode-hook (if diff-hl-mode 'diff-hl-mode-on-hook 'diff-hl-mode-off-hook)) (if (called-interactively-p 'any) (progn nil (if (and (current-message) (not (equal last-message (current-message)))) nil (let ((local " in current buffer")) (message "Diff-Hl mode %sabled%s" (if diff-hl-mode "en" "dis") local))))))
  diff-hl-mode(1)
  (cond (buffer-file-name (diff-hl-mode 1)) ((eq major-mode 'vc-dir-mode) (diff-hl-dir-mode 1)))
  turn-on-diff-hl-mode()
  (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode))
  (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode)))
  (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode))))
  (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode)))) (setq diff-hl-mode-major-mode major-mode))
  (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode)))) (setq diff-hl-mode-major-mode major-mode)))
  (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn (diff-hl-mode -1) (turn-on-diff-hl-mode)) (turn-on-diff-hl-mode)))) (setq diff-hl-mode-major-mode major-mode))))
  (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode (progn ... ...) (turn-on-diff-hl-mode)))) (setq diff-hl-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if (eq diff-hl-mode-major-mode major-mode) nil (if diff-hl-mode ... ...))) (setq diff-hl-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- buffers)) (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil (if ... nil ...)) (setq diff-hl-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (let ((buffers global-diff-hl-mode-buffers)) (setq global-diff-hl-mode-buffers nil) (let ((--dolist-tail-- buffers)) (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if diff-hl-mode-set-explicitly nil ...) (setq diff-hl-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  global-diff-hl-mode-enable-in-buffers()
  run-hooks(change-major-mode-after-body-hook after-change-major-mode-hook)
  normal-mode(t)
  after-find-file(nil t)
  find-file-noselect-1(#<buffer 1.txt> "~/1.txt" nil nil "~/1.txt" (19147035 66311))
  find-file-noselect("/home/magic/1.txt" nil nil nil)
  find-file("/home/magic/1.txt")
  helm-find-file-or-marked("/home/magic/1.txt")
  helm-execute-selection-action-1()
  helm-execute-selection-action()
  helm-internal(helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil)
  apply(helm-internal (helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil))
  helm(helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil)
  apply(helm (helm-source-find-files "/home/magic/" "Find files or url: " nil nil "*helm find files*" nil nil nil))
  helm(:sources helm-source-find-files :input "/home/magic/" :case-fold-search smart :preselect nil :ff-transformer-show-only-basename t :default nil :prompt "Find files or url: " :buffer "*helm find files*")
  helm-find-files-1("/home/magic/" nil)
  helm-find-files(nil)
  funcall-interactively(helm-find-files nil)
  call-interactively(helm-find-files nil nil)
  command-execute(helm-find-files)

```

**Comment by mgcyung** [comment]
I found this one works,
```diff
@@ -182,13 +182,15 @@
       )))
 
 (defun diff-hl-maybe-define-bitmaps ()
-  (when (window-system) ;; No fringes in the console.
+  (when (and (window-system)
+             (not (eql fringe-mode nil))) ;; No fringes in the console.
     (unless (fringe-bitmap-p 'diff-hl-bmp-empty)
       (diff-hl-define-bitmaps)
       (define-fringe-bitmap 'diff-hl-bmp-empty [0] 1 1 'center))))
 
 (defun diff-hl-maybe-redefine-bitmaps ()
-  (when (window-system)
+  (when (and (window-system)
+             (not (eql fringe-mode nil)))
     (diff-hl-define-bitmaps)))
 
 (defvar diff-hl-spec-cache (make-hash-table :test 'equal))
```

**Comment by dgutov** [comment]
OK, tell me this: how do you open Emacs?

Are you using a daemon? Terminal? Graphical session?

**Comment by mgcyung** [comment]
It's opened in GUI.

**Comment by dgutov** [comment]
And what do you do to make the problem happen? Disable the fringes?

**Comment by mgcyung** [comment]
Upgrade from 27.0 to 28.0

**Comment by dgutov** [comment]
I can't reproduce using Emacs 28 either (the current tip).

Please provide a recipe starting with `emacs -Q -L ~/.emacs.d/elpa/diff-hl-xxx -l diff-hl'.

**Comment by dgutov** [comment]
Let me know if this didn't fix the problem.

**Comment by saolaolsson** [comment]
The version available on ELPA/MELPA does not seem to include 2cf8b48 yet, so I encountered this exact issue after an upgrade to Emacs 27.1 (the one that recently appeared in Debian testing) from 26.3.

Including 2cf8b48 by hand fixed the issue. Not having fringe widths set to zero by default also fixed the issue.

**Comment by dgutov** [comment]
The version in MELPA surely includes that commit (it builds from master), but perhaps you meant MELPA Stable.

Thanks for the reminder, though.
